lib/src/wallet/traits.rs

```diff
@@ -0,0 +1,464 @@
+use super::{
+    data::{OrchardNoteData, SaplingNoteData, WalletNullifier, WalletTx, WitnessCache},
+    keys::{orchard::OrchardKey, sapling::SaplingKey},
+};
+use nonempty::NonEmpty;
+use orchard::{
+    bundle::{Authorization as OrchardAuthorization, Bundle as OrchardBundle},
+    keys::{
+        Diversifier as OrchardDiversifier, FullViewingKey as OrchardFullViewingKey,
+        IncomingViewingKey as OrchardIncomingViewingKey,
+        OutgoingViewingKey as OrchardOutgoingViewingKey, SpendingKey as OrchardSpendingKey,
+    },
+    note::{Note as OrchardNote, Nullifier as OrchardNullifier},
+    note_encryption::OrchardDomain,
+    tree::MerkleHashOrchard,
+    Action, Address as OrchardAddress,
+};
+use zcash_address::unified::{self, Encoding as _, Receiver};
+use zcash_client_backend::encoding::encode_payment_address;
+use zcash_note_encryption::{Domain, ShieldedOutput, ENC_CIPHERTEXT_SIZE};
+use zcash_primitives::{
+    consensus::Parameters,
+    keys::OutgoingViewingKey as SaplingOutgoingViewingKey,
+    memo::{Memo, MemoBytes},
+    merkle_tree::Hashable,
+    sapling::{
+        note_encryption::SaplingDomain, Diversifier as SaplingDiversifier, Node as SaplingNode,
+        Note as SaplingNote, Nullifier as SaplingNullifier, PaymentAddress as SaplingAddress,
+        SaplingIvk,
+    },
+    transaction::{
+        components::{
+            sapling::{
+                Authorization as SaplingAuthorization, Bundle as SaplingBundle, GrothProofBytes,
+            },
+            OutputDescription, SpendDescription,
+        },
+        TxId,
+    },
+    zip32::{
+        ExtendedFullViewingKey as SaplingExtendedFullViewingKey,
+        ExtendedSpendingKey as SaplingExtendedSpendingKey,
+    },
+};
+use zingoconfig::Network;
+
+pub(crate) trait ToBytes<const N: usize> {
+    fn to_bytes(&self) -> [u8; N];
+}
+
+impl ToBytes<32> for SaplingNullifier {
+    fn to_bytes(&self) -> [u8; 32] {
+        self.0
+    }
+}
+
+impl ToBytes<32> for OrchardNullifier {
+    fn to_bytes(&self) -> [u8; 32] {
+        OrchardNullifier::to_bytes(*self)
+    }
+}
+
+impl ToBytes<512> for Memo {
+    fn to_bytes(&self) -> [u8; 512] {
+        *self.encode().as_array()
+    }
+}
+
+impl ToBytes<512> for MemoBytes {
+    fn to_bytes(&self) -> [u8; 512] {
+        *self.as_array()
+    }
+}
+
+impl<const N: usize> ToBytes<N> for [u8; N] {
```

Comment:
Huh....

Reply:
We only need this for a certain value, but there's no reason not to implement it for any N.